File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.go
func lengthOfLongestSubstring(s string) int {
    str := []rune(s)
    maxSub := 0

    for i := 0; i < len(str); i++ {

        seen := map[rune]bool{}
        subStr := []rune{}

        for j := i; j < len(str); j++ {
            if v := seen[str[j]]; v {
                break
            }

            seen[str[j]] = true
            subStr = append(subStr, str[j])
        }

        if len(subStr) > maxSub {
            maxSub = len(subStr)
        }
    }

    return maxSub
}

/*
func lengthOfLongestSubstring(s string) int {
    last := map[rune]int{}

    maxSub := 0
    left := 0

    for i, v := range s {
        if idx, ok := last[v]; !ok || idx < left {
            sub := i - left + 1
            if sub > maxSub {
                maxSub = sub
            }
        } else {
            left = last[v] + 1
        }

        last[v] = i
    }

    return maxSub
}
*/

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
"""
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        max_sub = 0
        for i in range(len(s)):
            sub = ""
            occ = {}
            for j in range(i, len(s)):
                if occ.get(s[j]):
                    break
                else:
                    sub += s[j]
                    occ[s[j]] = 1

                if j == len(s) - 1:
                    return max(len(sub), max_sub)

            max_sub = max(len(sub), max_sub)

        return max_sub

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        max_sub = 0
        for i in range(len(s)):
            sub = ""
            for j in range(i, len(s)):
                if s[j] in sub:
                    break
                else:
                    sub += s[j]

            max_sub = max(len(sub), max_sub)       

        return max_sub    
""" 

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        last = {}
        left = 0
        max_sub = 0

        for i in range(len(s)):
            if last.get(s[i], -1) < left:
                # either there's no duplicate, or the duplicate is irrelevant
                # because it occurs at a place less than the left index
                max_sub = max(i - left + 1, max_sub)    
            else:
                # maintain a left index greater than the last duplicate character
                left = last[s[i]] + 1

            last[s[i]] = i

        return max_sub    



